transport_plugins/bled112/iotile_transport_bled112/bled112.py

```diff
@@ -184,6 +186,38 @@ def reset_scan_stats(self):
         self._device_scan_counts = {}
         self._last_reset_time = time.monotonic()
 
+    def _update_conn_map(self, conn_string, device_uuid):
+        """
+        Updates the instance's mapping of device UUIDs to connection strings.
+
+        This uses a simple implementation of a Least Recently Used algorithm.
+        Everytime an entry is updated, it will remove the existing entry
+        and reinstert it at the end. As a result, the least recently used entries
+        will be towards the beginning of the list. That way if our list grows
+        too large and needs to be overwritten, it will do so by removing the
+        beginning elements.
+
+        Args:
+            conn_string (str): The string of the device's MAC address
+            device_uuid (int): The integer representation of the device's UUID
+        """
+        entry = (conn_string, device_uuid)
+
+        if entry in self._conn_map:
+            self._conn_map.remove(entry)
+        elif len(self._conn_map) >= self.ConnMapMaxSize:
+            self._conn_map = self._conn_map[-(self.ConnMapMaxSize-1):]
+
+        self._conn_map.append(entry)
+
+    def _get_device_alias(self, device_id):
```

Comment:
I don't think you should ever need a bidirectional lookup to implement this feature.  If you do, then there's a bug somewhere else in the existing implementation that should be fixed instead of working around it with a bidirectional lookup.

Can you verify:
1. `context['uuid']` is only ever used in the authentication related functions.  If not, lets discuss where else it is used.  The below assumes it was added previously to enable this feature.

Given the above, the underlying problem is that `context['uuid']` is incorrectly set in the code.  It should have always been the `device_id` not the mac address/connection string.  The fix is to do the `connection_string -> device_id` lookup before setting `context['uuid']` and just set it to `device_id` if known and `None` if the lookup fails.

During authentication then, `context['uuid']` is always either the correct device_id or a `None`.  If auth is required and `context['uuid'] is None` then auth fails, otherwise it proceeds.

There should never need to be a lookup of connection string from `device_id`.  Let me know if I'm missing something in the code.  Everywhere inside of bled112 device adapter knows the connection string already, it's just the device id that the code doesn't know.

Reply:
Got it. I thought it was a bit odd and wondered if we should just drop using the MAC as a potential look up. I looked through and tested the code and verified the `context['uuid'] is only used in those authentication functions. I went ahead and fixed where that `connection_string` was initially fed into the chain of callbacks (inside of `_probe_characteristics_finished`).
I also am just using the standard `get` function on the dictionary. That will automatically return `None` if it doesn't exist, which will get caught inside the check that I added in the `BLED112AuthManager`.